### ISSUE TITLE
Display of non-fundamental cell types

### ISSUE THREAD

**Issue by nokome** [stencila/stencila#100]
Currently, cells which have a `tipe` that is not fundamental (ie. `real`, `integer` etc) are represented with an indicator of their type and their string representation value which is scollable: e.g.

![image](https://cloud.githubusercontent.com/assets/1152336/12283918/1ca5394e-ba10-11e5-94e4-f6d820f1b5c1.png)

But this can mean the cells take up a lot of space. it may be better to represent them as a standard size cell and allow expansion of their contents in a side panel.

**Comment by michael** [comment]
So you don't think it's important to be able to read the contents of such cells in full?

Somewhat related: When I watched your video I saw that you placed a new plot and when it rendered it was displayed over existing content. We could dynamically determine the height of a row depending on the tallest cell of that row. However then it would look a bit ugly for fundamental cell types that only have a number it and get a lot of vertical spacing.

**Comment by nokome** [comment]
@michael : yes, it's important to be able to read the contents of such cells in full - it would be really frustrating if you could not do that somewhere - but where is the best place? - my suggestion is the side panel.

Yes, the placement of plots is very much related. Originally with the plots I had them within the cells (just as you say except CSS does everything for you - no need to dynamically determine heights etc). But that was ugly so I added this:

```css
td.display.ImageFile {
  position: relative;
  img {
    position: absolute;
    top: 3px;
    left: 3px;
    border: 1px solid #EEE;
    z-index: 1;
  }
}
```

to replicate behaviour in Excel etc. We need to add dynamic resizing of plots somehow too!

**Comment by michael** [comment]
We could solve this inline as an overlay. 

![image](https://cloud.githubusercontent.com/assets/2931/12392179/53955854-bded-11e5-9493-464351e3c775.png)

The overlay would expand fully, so there are no additional scrollbars. We'll try out how that works. After some thinking I'm for trying to avoid the sidebar and rather use horizontal space for content. Let's see how far we get with displaying contextual information inline.

**Comment by michael** [comment]
We experimented with cell expanding/collapsing that the user can toggle using the spacebar when on a cell. Additionally we could provide + or - icons for the clickers.

![image](https://cloud.githubusercontent.com/assets/2931/12398800/38dc4256-be16-11e5-998f-daa9797d53e2.png)

I didn't feel too wrong to just expand the cell and let the cell grow instead of an overlay. The advantage is that you never cover other content. We could do the same with plots.

**Comment by nokome** [comment]
Great! This looks really good. I'll have a play with it an let you know if I have any further suggestions.

**Comment by nokome** [comment]
I'm not seeing the above yet when running in devmode  - perhaps it's not yet pushed - or the above above just mockups! Anyway...

As well as having the +/- perhaps we need a "pin" for the content to stay open regardless of who views the sheet - would need to get persisted in 'layout.json' (#101) or somewhere else.

I think overlay rather than cell expansion is the way to go - it would just get too ugly otherwise (lots of whitespace). With the +/- it does not matter if other content is covered. For plots, overlay is what ppl are used to in Excel etc

**Comment by nokome** [comment]
Images R produces are now 800x800px - arbitrary and temporary. Will think about using `aspect` and `quality` specs to customize this. Also resizing of images.

**Comment by michael** [comment]
I implemented the tristate toggling of the displayMode yesterday, as shown in the screenshot. However while toggling with the space key works nicely, the mouse interaction is crappy with that extra toggle UI element. Because the click surface conflicts with the drag behavior to select cells. Probably we could solve this with some state keeping of the drag state but since I'm also not happy visually I would suggest we try one of those options:

- we move that interaction into the toolbar (DisplayModeTool)
- context menu (right click -> change display mode -> overlay)

Or both. What would you prefer?

**Comment by nokome** [comment]
I have not tried it out yet. But from what you say, it sounds like it would be better to put it in the toolbar rather than introduce a context menu.

**Comment by michael** [comment]
Looks like this atm. 

![toggle-changemode](https://cloud.githubusercontent.com/assets/2931/12493396/28f831bc-c085-11e5-854b-c65e231d9e27.gif)

As you see it covers another cell's content. I think toolbar version will be better and more explicit as we can provide a dropdown with the options in plain english rather than icons.

**Comment by nokome** [comment]
Why not just put the blue toggle inside the cell itself - so it does not cover anyone else. Also make it partially transparent. Having not tried it, I might be missing something. Will try now.

**Comment by nokome** [comment]
OK tried it now. I was wondering why images were not coming up yesterday - you need to keyboard over the cell to get the toggle! Now I understand it - nice! :tada: Some suggestions:

- having tried it now, I think it would be a shame to loose that keyboard interaction and the inline toggle. But we really do need to fix up mouse interaction. Now, I'm not so sure about what you are thinking about putting in the toolbar - the toggle?
- the toggle should always be shown - within the cell it relates to
- default to minimized as now, except images which should default to overlay

**Comment by michael** [comment]
We had a bunch of technical difficulties with the mouse interaction. E.g. it broke double clicking a cell for editing. I disabled the fragile code now and things work faster and more reliable.

However we can bring that toggle back with a smarter implementation and displaying the toggle inside the cell. But maybe it's smart to skip it for this iteration for stability sake.... Anyways.. let's talk tomorrow morning... say 9:30?

**Comment by nokome** [comment]
OK, thanks
